# 🌾 Notebook 2: RAG Pipeline
### Agriculture Voice Assistant — Pakistan Farmers

---

## Pipeline Flow
```
Farmer Voice Message (Urdu/Punjabi audio)
         ↓
[STT] Fine-tuned Whisper  →  Roman Urdu/Punjabi text
         ↓
[Step 1] LLM Query Enhancement  →  Qwen2.5 (Ollama, local, free)
         Understands Roman Urdu/Punjabi
         Extracts: crop, topic, stage, intent, keywords
         Cleans + structures the question before going into RAG
         ↓
[Step 2] Vector Search  →  ChromaDB (built by Notebook 1)
         Returns top-5 relevant Q&A pairs
         PRINTS: Raw best-match answer from knowledge base
         ↓
[Step 3] LLM Refinement  →  Qwen2.5 (Ollama)
         Uses RAG context to produce a natural reply
         Replies in same language as the farmer
         PRINTS: Final refined answer
```

---
**Prerequisite:** Run Notebook 1 first to build the ChromaDB knowledge base.

## Step 0 — Install Ollama & Pull Model
First-time only. Downloads ~4.7GB for Qwen2.5:7b.

In [1]:
# import subprocess, time, os

# def run(cmd):
#     result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
#     if result.returncode != 0:
#         print(f"STDERR: {result.stderr[:300]}")
#     return result.stdout.strip()

# # Uncomment all lines below to install on Linux/Colab:
# # print("Installing Ollama...")
# # os.system("curl -fsSL https://ollama.com/install.sh | sh")
# # print("Starting Ollama server...")
# # subprocess.Popen(["ollama", "serve"], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
# # time.sleep(4)
# # print("Pulling qwen2.5:7b...")
# # os.system("ollama pull qwen2.5:7b")

# models = run("ollama list")
# print("✅ Available Ollama models:")
# print(models)

## Step 1 — Install Python Dependencies

In [2]:
# !pip install -q sentence-transformers chromadb requests pandas transformers torch torchaudio edge-tts nest_asyncio
# print("✅ Dependencies installed.")

## Step 2 — Configuration

In [3]:
# ============================================================
# Configuration

# ChromaDB path — must match Notebook 1
CHROMA_DB_PATH  = "./content/agriculture_chroma_db"
COLLECTION_NAME = "agriculture_kb"

# Embedding model — must be same as Notebook 1
EMBEDDING_MODEL = "all-MiniLM-L6-v2"

# Ollama
OLLAMA_URL   = "http://localhost:11434/api/generate"
OLLAMA_MODEL = "gemma2:9b" # use gemma2:9b(best i think) or qwen2.5:7b or qwen3:8b (heavier than the other two)

# Your fine-tuned Whisper Urdu v2 model path
WHISPER_MODEL_PATH = "../stt-finetune/models/whisper_urdu_finetuned_v2"
# Optional: set to "ur", "en", etc. to force a language; None = auto
WHISPER_LANGUAGE   = None

print("Configuration loaded.")
print(f"  ChromaDB    : {CHROMA_DB_PATH}")
print(f"  Ollama model: {OLLAMA_MODEL}")
print(f"  Whisper     : {WHISPER_MODEL_PATH}")
print(f"  Whisper lang: {WHISPER_LANGUAGE or 'auto'}")

Configuration loaded.
  ChromaDB    : ./content/agriculture_chroma_db
  Ollama model: gemma2:9b
  Whisper     : ../stt-finetune/models/whisper_urdu_finetuned_v2
  Whisper lang: auto


## Step 3 — Import Libraries

In [ ]:
import json
import requests
import time
import os
import warnings
import torch
import chromadb
import pandas as pd
from sentence_transformers import SentenceTransformer
from typing import Dict, List
from transformers import pipeline as hf_pipeline
import asyncio
from datetime import datetime
import edge_tts
warnings.filterwarnings('ignore')

print("✅ Libraries imported.")
print(f"   PyTorch  : {torch.__version__}")
print(f"   CUDA     : {torch.cuda.is_available()}")
if torch.cuda.is_available():
    print(f"   GPU      : {torch.cuda.get_device_name(0)}")
    print(f"   VRAM     : {torch.cuda.get_device_properties(0).total_memory / 1e9:.1f} GB")


✅ Libraries imported.
   PyTorch  : 2.5.1+cu121
   CUDA     : True
   GPU      : NVIDIA GeForce RTX 4060 Laptop GPU
   VRAM     : 8.6 GB


## Step 4 — Load ChromaDB, Embedding Model & Verify Ollama

In [5]:
# ── ChromaDB ───────────────────────────────────────────────
print("Loading ChromaDB knowledge base...")
chroma_client = chromadb.PersistentClient(path=CHROMA_DB_PATH)
collection    = chroma_client.get_collection(name=COLLECTION_NAME)
print(f"✅ ChromaDB loaded — {collection.count()} documents indexed.")

# ── Embedding Model ────────────────────────────────────────
print("\nLoading embedding model...")
embedding_model = SentenceTransformer(EMBEDDING_MODEL)
print(f"✅ Embedding model ready: {EMBEDDING_MODEL}")

# ── Device ─────────────────────────────────────────────────
device = "cuda" if torch.cuda.is_available() else "cpu"
print(f"\n✅ Device: {device}")
if torch.cuda.is_available():
    print(f"   GPU  : {torch.cuda.get_device_name(0)}")
    print(f"   VRAM : {torch.cuda.get_device_properties(0).total_memory / 1e9:.1f} GB")

# ── Ollama ─────────────────────────────────────────────────
print("\nChecking Ollama server...")
try:
    r = requests.get("http://localhost:11434/api/tags", timeout=5)
    models_available = [m['name'] for m in r.json().get('models', [])]
    if any(OLLAMA_MODEL in m for m in models_available):
        print(f"✅ Ollama running. Model '{OLLAMA_MODEL}' available.")
    else:
        print(f"⚠️  '{OLLAMA_MODEL}' not found. Run: ollama pull {OLLAMA_MODEL}")
        print(f"   Available: {models_available}")
except Exception as e:
    print(f"❌ Ollama not reachable: {e}")
    print("   Start with: !ollama serve &")

print("\n" + "=" * 60)
print("ALL LOADED — Ready.")
print("=" * 60)

Loading ChromaDB knowledge base...
✅ ChromaDB loaded — 3112 documents indexed.

Loading embedding model...
✅ Embedding model ready: all-MiniLM-L6-v2

✅ Device: cuda
   GPU  : NVIDIA GeForce RTX 4060 Laptop GPU
   VRAM : 8.6 GB

Checking Ollama server...
✅ Ollama running. Model 'gemma2:9b' available.

ALL LOADED — Ready.


## Step 5 — STT: Fine-tuned Whisper Urdu v2
Loads your fine-tuned Whisper and transcribes farmer audio (output format depends on the model).

In [28]:
import subprocess
import numpy as np
from pathlib import Path
from transformers import (
    WhisperProcessor,
    WhisperForConditionalGeneration,
    WhisperFeatureExtractor,
    WhisperTokenizer,
 )

MODEL_PATH = Path(WHISPER_MODEL_PATH)

ALLOWED_AUDIO_EXTS = {".wav", ".mp3", ".m4a", ".ogg", ".flac", ".aac", ".wma", ".opus"}


def validate_audio_path(audio_path: str) -> str:
    if not audio_path or not str(audio_path).strip():
        raise ValueError("audio_path is empty.")
    normalized = os.path.expanduser(str(audio_path).strip())
    if not os.path.exists(normalized):
        raise FileNotFoundError(f"Audio file not found: {normalized}")
    ext = Path(normalized).suffix.lower()
    if ext not in ALLOWED_AUDIO_EXTS:
        allowed = ", ".join(sorted(ALLOWED_AUDIO_EXTS))
        raise ValueError(f"Unsupported audio extension '{ext}'. Allowed: {allowed}")
    return normalized


def load_audio_ffmpeg(path, target_sr=16000):
    """Load any audio file via FFmpeg — handles .wav, .mp3, .m4a, .ogg, etc."""
    cmd = [
        "ffmpeg", "-i", str(path),
        "-f", "f32le", "-ac", "1", "-ar", str(target_sr), "-"
    ]
    out = subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=True)
    audio = np.frombuffer(out.stdout, np.float32)
    if audio.size == 0:
        raise RuntimeError(
            f"No audio decoded from {path}.\n{out.stderr.decode(errors='ignore')}"
        )
    return audio


def load_whisper_processor(model_path: Path) -> WhisperProcessor:
    """Load processor with fallback for checkpoints missing preprocessor_config.json."""
    if (model_path / "preprocessor_config.json").exists():
        return WhisperProcessor.from_pretrained(str(model_path))
    feature_extractor = WhisperFeatureExtractor(
        feature_size=128, sampling_rate=16000, hop_length=160,
        chunk_length=30, n_fft=400, padding_value=0.0,
        do_normalize=True, return_attention_mask=False,
    )
    has_local_tokenizer = (
        (model_path / "vocab.json").exists() and
        (model_path / "merges.txt").exists()
    )
    tokenizer = (
        WhisperTokenizer.from_pretrained(str(model_path))
        if has_local_tokenizer
        else WhisperTokenizer.from_pretrained("openai/whisper-large-v3")
    )
    return WhisperProcessor(feature_extractor=feature_extractor, tokenizer=tokenizer)


# Load model once at startup
print(f"Loading fine-tuned Whisper from: {MODEL_PATH}")
whisper_processor = load_whisper_processor(MODEL_PATH)
whisper_model     = WhisperForConditionalGeneration.from_pretrained(str(MODEL_PATH))
whisper_model.to(device).eval()
print("✅ Fine-tuned Whisper loaded.")


def transcribe_audio(audio_path: str, verbose: bool = True) -> str:
    """
    Transcribe farmer audio using fine-tuned Whisper Urdu v2.
    Output format depends on the model (no forced language by default).
    """
    audio_path = validate_audio_path(audio_path)
    if verbose:
        print("=" * 60)
        print("STT — SPEECH TO TEXT (Fine-tuned Whisper Urdu v2)")
        print("=" * 60)
        print(f"Audio: {audio_path}")

    audio = load_audio_ffmpeg(audio_path, target_sr=16000)

    # inputs = whisper_processor(audio, sampling_rate=16000, return_tensors="pt")
    # input_features = inputs.input_features.to(
    #     device=device,
    #     dtype=next(whisper_model.parameters()).dtype
    # )

    # generate_kwargs = {"num_beams": 2, "task": "transcribe"}
    # if WHISPER_LANGUAGE:
    #     generate_kwargs["language"] = WHISPER_LANGUAGE

    # with torch.no_grad():
    #     predicted_ids = whisper_model.generate(
    #         input_features,
    #         **generate_kwargs
    #     )

    # transcription = whisper_processor.tokenizer.decode(
    #     predicted_ids[0], skip_special_tokens=True
    # ).strip()

    CHUNK_S  = 28        # seconds per chunk (keep under 30)
    STRIDE_S = 2         # overlap so words at boundaries aren't lost
    SR       = 16000
    chunk_samples  = CHUNK_S  * SR
    stride_samples = STRIDE_S * SR

    generate_kwargs = {"num_beams": 2, "task": "transcribe"}
    if WHISPER_LANGUAGE:
        generate_kwargs["language"] = WHISPER_LANGUAGE

    dtype = next(whisper_model.parameters()).dtype
    parts = []

    start = 0
    while start < len(audio):
        chunk = audio[start : start + chunk_samples]
        inputs = whisper_processor(chunk, sampling_rate=SR, return_tensors="pt")
        input_features = inputs.input_features.to(device=device, dtype=dtype)
        with torch.no_grad():
            predicted_ids = whisper_model.generate(input_features, **generate_kwargs)
        part = whisper_processor.tokenizer.decode(
            predicted_ids[0], skip_special_tokens=True
        ).strip()
        if part:
            parts.append(part)
        # advance by chunk minus overlap so next chunk re-covers the last STRIDE_S seconds
        start += chunk_samples - stride_samples

    transcription = " ".join(parts)
    if verbose:
        print(f"\n✅ Transcription: {transcription}")

    return transcription

Loading fine-tuned Whisper from: ..\stt-finetune\models\whisper_urdu_finetuned_v2
✅ Fine-tuned Whisper loaded.


## Step 6 — LLM Query Enhancement (Qwen2.5 via Ollama)
Takes raw Roman Urdu/Punjabi → translates + extracts structured metadata.  
This cleaned, English version is what actually goes into ChromaDB search.

In [29]:
def call_ollama(prompt: str, system_prompt: str = "", temperature: float = 0.1) -> str:
    """Call Qwen2.5:7b via Ollama local API."""
    full_prompt = f"{system_prompt}\n\n{prompt}" if system_prompt else prompt
    payload = {
        "model"  : OLLAMA_MODEL,
        "prompt" : full_prompt,
        "stream" : False,
        "options": {"temperature": temperature, "num_predict": 512}
    }
    response = requests.post(OLLAMA_URL, json=payload, timeout=120)
    response.raise_for_status()
    return response.json()["response"].strip()


def enhance_farmer_query(raw_question: str, verbose: bool = True) -> Dict:
    """
    LLM Step 1: Faithful translation + metadata extraction for better RAG retrieval.
    Priority order:
    1) Preserve meaning exactly
    2) Produce clean English query for embedding search
    3) Extract conservative metadata (avoid guessing)
    """

    system_prompt = """You are a translation-first agricultural query normalizer for Pakistani farmer questions.
You understand Roman Urdu, Roman Punjabi, Urdu terms written in Latin script and Urdu/Arabic script, and English.

NON-NEGOTIABLE RULES:
- Translate faithfully. Do not change crop, input, disease, pest, timing, quantity, or intent.
- Never substitute one crop/input for another.
- If a source term is ambiguous, keep it in parentheses in enhanced_query.
- Do not invent missing details. Use Unknown or Not specified when unclear.
- Prefer literal correctness over fluent rewriting.
- Output STRICT JSON only (no markdown, no extra text)."""

    enhancement_prompt = f"""Farmer question: "{raw_question}"

Return EXACTLY one JSON object with these keys:
1. "enhanced_query": precise English translation used for retrieval
2. "detected_language": one of ["Roman Urdu", "Roman Punjabi", "Urdu", "Punjabi", "English", "Mixed"]
3. "crop": one of common crops or "Unknown"
4. "topic": one of [Sowing, Seed Rate, Nursery, Transplanting, Fertilizer, Irrigation, Weed Management, Pest Management, Disease Management, Harvesting, Yield, Variety, Soil, Climate, Land Preparation, General]
5. "stage": one of [Pre-Sowing, Nursery, Germination, Vegetative, Flowering, Fruit Formation, Grain Formation, Maturity, Harvest, Post-Harvest, Any]
6. "intent_type": one of [Recommendation, Prevention, Chemical Control, Biological Control, Cultural Control, Mechanical Control, Symptoms, Impact, Duration, Identification, Dosage, Timing, Fact, Comparison]
7. "entity": specific disease/pest/fertilizer/practice or "Not specified"
8. "keywords": list of 3-5 short English search keywords, lowercase, no duplicates
9. "season": one of [Kharif, Rabi, Spring, Summer, Winter, Not specified]
10. "reply_language": always "English"
11. "translation_confidence": number between 0 and 1
12. "ambiguity_notes": short string; "None" if clear

Important: if uncertain about metadata fields, be conservative and use Unknown/Not specified rather than guessing.
Return JSON only."""

    try:
        raw_response = call_ollama(enhancement_prompt, system_prompt, temperature=0.1)

        # Strip markdown fences if model wrapped the JSON
        if "```json" in raw_response:
            raw_response = raw_response.split("```json")[1].split("```")[0]
        elif "```" in raw_response:
            raw_response = raw_response.split("```")[1].split("```")[0]

        enhanced = json.loads(raw_response.strip())

        # Defensive normalization to keep downstream code stable
        defaults = {
            "enhanced_query": raw_question,
            "detected_language": "Unknown",
            "crop": "Unknown",
            "topic": "General",
            "stage": "Any",
            "intent_type": "Fact",
            "entity": "Not specified",
            "keywords": raw_question.split()[:5],
            "season": "Not specified",
            "reply_language": "English",
            "translation_confidence": 0.5,
            "ambiguity_notes": "None",
        }
        for k, v in defaults.items():
            if k not in enhanced or enhanced[k] in (None, ""):
                enhanced[k] = v

        if not isinstance(enhanced.get("keywords"), list):
            enhanced["keywords"] = str(enhanced["keywords"]).split()[:5]

        # Normalize keywords: lowercase unique strings, max 5
        normalized_keywords = []
        for kw in enhanced.get("keywords", []):
            kw_s = str(kw).strip().lower()
            if kw_s and kw_s not in normalized_keywords:
                normalized_keywords.append(kw_s)
        enhanced["keywords"] = normalized_keywords[:5]

        if verbose:
            print("\n" + "=" * 60)
            print(f"STEP 1 — LLM QUERY ENHANCEMENT ({OLLAMA_MODEL})")
            print("=" * 60)
            print(f"  Original          : {raw_question}")
            print(f"  Detected language : {enhanced.get('detected_language', '?')}")
            print(f"  Enhanced (English): {enhanced.get('enhanced_query', raw_question)}")
            print(f"  Crop              : {enhanced.get('crop', 'Unknown')}")
            print(f"  Topic             : {enhanced.get('topic', 'General')}")
            print(f"  Stage             : {enhanced.get('stage', 'Any')}")
            print(f"  Intent            : {enhanced.get('intent_type', 'Fact')}")
            print(f"  Entity            : {enhanced.get('entity', 'Not specified')}")
            print(f"  Keywords          : {', '.join(enhanced.get('keywords', []))}")
            print(f"  Translation conf  : {enhanced.get('translation_confidence', 0.5)}")
            print(f"  Ambiguity notes   : {enhanced.get('ambiguity_notes', 'None')}")
            print("=" * 60)

        return enhanced

    except (json.JSONDecodeError, Exception) as e:
        print(f"WARNING: Enhancement failed ({e}), using fallback.")
        return {
            "enhanced_query"        : raw_question,
            "detected_language"     : "Unknown",
            "crop"                  : "Unknown",
            "topic"                 : "General",
            "stage"                 : "Any",
            "intent_type"           : "Fact",
            "entity"                : "Not specified",
            "keywords"              : raw_question.split()[:5],
            "season"                : "Not specified",
            "reply_language"        : "English",
            "translation_confidence": 0.5,
            "ambiguity_notes"       : "Fallback mode",
        }


print("OK: enhance_farmer_query() defined.")

OK: enhance_farmer_query() defined.


## Step 7 — RAG Vector Search (ChromaDB)
Searches the knowledge base using the LLM-enhanced query.

In [30]:
def rag_search(enhanced_query: Dict, top_k: int = 5, verbose: bool = True) -> List[Dict]:
    """
    Vector similarity search in ChromaDB.
    - Embeds enhanced English query + keywords
    - Applies crop+topic metadata filter if both are known
    - Falls back to no filter if filtered search fails
    """

    search_text = enhanced_query.get('enhanced_query', '')
    if enhanced_query.get('keywords'):
        search_text += " " + " ".join(enhanced_query['keywords'])
    if enhanced_query.get('entity') and enhanced_query['entity'] != 'Not specified':
        search_text += " " + enhanced_query['entity']

    query_embedding = embedding_model.encode(search_text).tolist()

    crop  = enhanced_query.get('crop', 'Unknown')
    topic = enhanced_query.get('topic', 'General')

    where_filter = {}
    if crop not in ('Unknown', 'Not specified', '') and \
       topic not in ('General', 'Not specified', ''):
        where_filter = {"$and": [
            {"crop" : {"$eq": crop}},
            {"topic": {"$eq": topic}}
        ]}
    elif crop not in ('Unknown', 'Not specified', ''):
        where_filter = {"crop": {"$eq": crop}}

    if verbose:
        print("\n" + "=" * 60)
        print("STEP 2 — RAG VECTOR SEARCH (ChromaDB)")
        print("=" * 60)
        print(f"  Search text : {search_text[:100]}")
        print(f"  Filter      : {where_filter if where_filter else 'None (searching all)'}")

    try:
        results = collection.query(
            query_embeddings=[query_embedding],
            n_results=top_k,
            where=where_filter if where_filter else None
        )
    except Exception as e:
        print(f"  ⚠️  Filtered search failed ({e}), retrying without filter...")
        results = collection.query(
            query_embeddings=[query_embedding],
            n_results=top_k
        )

    formatted = []
    if results and results['metadatas']:
        for i, meta in enumerate(results['metadatas'][0]):
            formatted.append({
                'question'   : meta['question'],
                'answer'     : meta['answer'],
                'crop'       : meta['crop'],
                'topic'      : meta['topic'],
                'stage'      : meta['stage'],
                'intent_type': meta['intent_type'],
                'entity'     : meta['entity'],
                'similarity' : round(1 - results['distances'][0][i], 4)
            })

    if verbose:
        print(f"\n  Found {len(formatted)} results:")
        for i, r in enumerate(formatted[:3]):
            print(f"\n  Result #{i+1} (similarity={r['similarity']:.3f}):")
            print(f"    Q: {r['question'][:80]}")
            print(f"    A: {r['answer'][:120]}...")
            print(f"    Crop={r['crop']}, Topic={r['topic']}, Stage={r['stage']}")
        print("=" * 60)

    return formatted


print("✅ rag_search() defined.")

✅ rag_search() defined.


## Step 8 — LLM Response Generation (Qwen2.5 via Ollama)
Refines the raw RAG answer into a natural farmer-friendly response in the farmer's own language.

In [49]:
def generate_farmer_response(original_question: str,
                              rag_results: List[Dict],
                              enhanced_query: Dict,
                              verbose: bool = True) -> Dict:
    """
    Two-stage response:
      Stage A — raw_rag_answer: best-match answer pulled directly from knowledge base
      Stage B — refined_answer: Qwen2.5 refines it into natural farmer language
    Returns both as a dict.
    """

    reply_language = enhanced_query.get('reply_language', 'English')
    detected_lang  = enhanced_query.get('detected_language', 'English')

    lang_instructions = {
        "Roman Urdu"   : "Reply in Roman Urdu (Urdu written in English letters). "
                         "Example: 'Aapko gehuun par DAP khad dalni chahiye'.",
        "Roman Punjabi": "Reply in Roman Punjabi (Punjabi written in English letters). "
                         "Example: 'Tenu kapah nu paani dena chahida hai'.",
        "English"      : "Reply in clear, simple English.",
        "Mixed"        : "Reply in simple Roman Urdu mixed with English agricultural terms."
    }

    lang_instruction = "Reply in clear, simple Urdu script."

    # ── Stage A: Pull raw RAG answer ────────────────────────
    if rag_results:
        raw_rag_answer = rag_results[0]['answer']
        context_parts  = []
        for i, r in enumerate(rag_results[:3]):
            context_parts.append(
                f"Source {i+1} (Crop: {r['crop']}, Topic: {r['topic']}, "
                f"Similarity: {r['similarity']}):\n"
                f"Q: {r['question']}\nA: {r['answer']}"
            )
        context = "\n\n".join(context_parts)
        kb_instruction = "Use the knowledge base information above to answer."
    else:
        # Pure LLM fallback — no RAG context
        raw_rag_answer = None
        context        = "No relevant information was found in the knowledge base."
        kb_instruction = (
            "No knowledge base match was found. Answer from your own agricultural knowledge. "
            "Be honest if you are not certain."
        )

    # ── Stage B: LLM refines the answer ────────────────────
    system_prompt = """You are a helpful agricultural advisor for Pakistani farmers.
You are warm, practical, and speak like a knowledgeable local expert.
Always reply in simple, easy Urdu script that a rural farmer can understand.
For chemical names, dosages, and numbers, always write them in English as they are — do not translate them.
Example: "گندم کی زنگ کے لیے Propiconazole 1ml فی لیٹر پانی میں ملا کر spray کریں۔" """

    response_prompt = f"""A Pakistani farmer asked: "{original_question}"
Intent: {enhanced_query.get('intent_type', 'Recommendation')}

Knowledge base context:
{context}

Rules:
- Reply in simple, easy Urdu script. Use plain everyday Urdu a farmer would understand, not formal or difficult words.
- Chemical names, medicine names, dosages, and numbers must stay in English exactly as they are (e.g. "DAP 50kg", "Chlorpyrifos 2.5ml").
- 1-2 sentences MAXIMUM
- Start directly with the answer, no preamble
- Only include: what to do, what chemical/dose if available
- IMPORTANT: Read the context carefully. If a piece of information describes CONDITIONS that cause or spread a disease, do NOT present it as treatment timing or application advice.
- If the KB does not contain specific treatment info, give advice if you are aware of the thing being talked about.
- No filler, no restating the question

Answer:"""

    if verbose:
        print("\n" + "=" * 60)
        print(f"STEP 3 — LLM RESPONSE GENERATION ({OLLAMA_MODEL})")
        print("=" * 60)
        print(f"  Reply language : {reply_language}")
        print(f"  RAG sources    : {len(rag_results)} documents")
        print("  Generating...")

    try:
        refined_answer = call_ollama(response_prompt, system_prompt, temperature=0.3)
        if verbose:
            print("=" * 60)
        return {
            "raw_rag_answer": raw_rag_answer,
            "refined_answer": refined_answer,
        }
    except Exception as e:
        print(f"❌ LLM failed ({e}) — returning raw RAG answer.")
        return {
            "raw_rag_answer": raw_rag_answer,
            "refined_answer": raw_rag_answer,
        }


print("✅ generate_farmer_response() defined.")

✅ generate_farmer_response() defined.


## Step 9 — Text-To-Speech


In [50]:
import requests
import base64
import os
from datetime import datetime

GOOGLE_TTS_API_KEY = "AIzaSyAMDBsup8S7rqexFcxSAO7tGSoumCfUSTo"   # ← paste your key

def text_to_speech_urdu(text: str,
                         output_path: str = None,
                         verbose: bool = True) -> str:
    """
    Convert English or Urdu text → natural Urdu speech using Google Cloud TTS.
    Saves as .mp3 and returns the file path.

    Voice: ur-PK-Wavenet-A  (female, natural Urdu)
    Alternatives:
        ur-PK-Wavenet-B   — male Urdu
        ur-PK-Standard-A  — female (less natural, but cheaper tier)
        ur-PK-Standard-B  — male standard
    """

    if output_path is None:
        os.makedirs("./audio_responses", exist_ok=True)
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        output_path = f"./audio_responses/response_{timestamp}.mp3"

    url = f"https://texttospeech.googleapis.com/v1/text:synthesize?key={GOOGLE_TTS_API_KEY}"

    payload = {
        "input": {
            "text": text
        },
        "voice": {
            "languageCode": "ur-IN",
            "name": "ur-IN-Chirp3-HD-Aoede",      # natural female Urdu voice
            "ssmlGender": "FEMALE"
        },
        "audioConfig": {
            "audioEncoding": "MP3",
            "speakingRate": 0.9,            # slightly slower — easier for farmers to follow
            "pitch": 0.0,                   # 0 = default pitch
            "volumeGainDb": 0.0
        }
    }

    if verbose:
        print("\n" + "=" * 60)
        print("TTS — TEXT TO SPEECH (Google Cloud, Urdu WaveNet)")
        print("=" * 60)
        print(f"  Text    : {text[:100]}{'...' if len(text) > 100 else ''}")
        print(f"  Voice   : ur-PK-Wavenet-A (natural female Urdu)")
        print(f"  Saving  : {output_path}")

    response = requests.post(url, json=payload, timeout=30)

    if response.status_code != 200:
        raise RuntimeError(
            f"Google TTS API error {response.status_code}: {response.text}"
        )

    audio_content = response.json().get("audioContent")
    if not audio_content:
        raise RuntimeError("No audio content returned from Google TTS.")

    audio_bytes = base64.b64decode(audio_content)
    with open(output_path, "wb") as f:
        f.write(audio_bytes)

    if verbose:
        size_kb = len(audio_bytes) / 1024
        print(f"  ✅ Saved {size_kb:.1f} KB audio → {output_path}")

    return output_path


print("✅ text_to_speech_urdu() defined.")

✅ text_to_speech_urdu() defined.


## Step 10 — Full Pipeline
Ties everything together. Prints RAG answer then LLM answer.

In [51]:
def run_full_pipeline(audio_path: str = None,
                      text_input: str = None,
                      verbose: bool = True) -> Dict:

    if (audio_path and text_input) or (not audio_path and not text_input):
        raise ValueError("Provide exactly one of audio_path or text_input.")

    input_mode = "voice" if audio_path else "text"

    print("\n" + "#" * 60)
    print("# AGRICULTURE ASSISTANT — FULL PIPELINE")
    print("#" * 60)
    if verbose:
        print(f"[Input mode: {input_mode}]")
    start_total = time.time()

    # ── 1. STT ─────────────────────────────────────────────
    if audio_path:
        t0               = time.time()
        farmer_text      = transcribe_audio(audio_path, verbose=verbose)
        stt_time         = time.time() - t0
        transcribed_text = farmer_text
        print(f"\n[STT text]: {transcribed_text}")
    else:
        farmer_text      = str(text_input).strip() if text_input is not None else ""
        if not farmer_text:
            raise ValueError("text_input is empty.")
        stt_time         = 0
        transcribed_text = None
        if verbose:
            print(f"\n[STT skipped — text input]: {farmer_text}")


    # ── 2. LLM Query Enhancement ────────────────────────────
    t0             = time.time()
    enhanced_query = enhance_farmer_query(farmer_text, verbose=verbose)
    enhance_time   = time.time() - t0

    # ── 3. RAG Search ───────────────────────────────────────
    SIMILARITY_THRESHOLD = 0.55  # raised — below this means RAG didn't find anything useful

    t0          = time.time()
    rag_results = rag_search(enhanced_query, top_k=5, verbose=False)  # verbose=False, we print manually below
    search_time = time.time() - t0

    # Always show ALL RAG results with their similarity scores
    print("\n" + "=" * 60)
    print("RAG RESULTS (all matches + similarity scores)")
    if rag_results:
        for i, r in enumerate(rag_results):
            status = "✅" if r['similarity'] >= SIMILARITY_THRESHOLD else "❌"
            print(f"\n  {status} Result #{i+1} — similarity={r['similarity']:.3f}")
            print(f"     Crop={r['crop']}, Topic={r['topic']}, Stage={r['stage']}")
            print(f"     Q: {r['question'][:80]}")
            print(f"     A: {r['answer'][:120]}...")
    else:
        print("  No results returned from ChromaDB.")

    # Filter to only confident matches
    good_results = [r for r in rag_results if r['similarity'] >= SIMILARITY_THRESHOLD]

    # ── 4. Decide: RAG or LLM fallback ─────────────────────
    print("\n" + "=" * 60)
    if good_results:
        print(f"RAW RAG ANSWER — using top match (sim={good_results[0]['similarity']:.3f})")
        print(good_results[0]['answer'])
        using_rag = True
    else:
        best_sim = rag_results[0]['similarity'] if rag_results else 0
        print(f"⚠️  RAG SKIPPED — best similarity was {best_sim:.3f} (threshold={SIMILARITY_THRESHOLD})")
        print("   Falling back to pure LLM answer.")
        using_rag = False

    # ── 5. LLM Response ─────────────────────────────────────
    t0       = time.time()
    response = generate_farmer_response(
        farmer_text, good_results, enhanced_query, verbose=verbose
    )
    gen_time = time.time() - t0

    # ── 6. Final Answer ─────────────────────────────────────
    print("\n" + "=" * 60)
    mode = "RAG + LLM" if using_rag else "Pure LLM (no RAG)"
    print(f"✅ FINAL ANSWER ({mode})")
    print(response["refined_answer"])

    total_time = time.time() - start_total
    # print("\n" + "-" * 60)
    # print(f"⏱️  STT={stt_time:.1f}s | Enhance={enhance_time:.1f}s | "
    #       f"Search={search_time:.2f}s | Refine={gen_time:.1f}s | Total={total_time:.1f}s")
    # print(f"📚 RAG used: {using_rag} | Good matches: {len(good_results)}/{len(rag_results)}")
    # print("-" * 60)

    # ── 7. TTS — Convert answer to Urdu speech ──────────────
    tts_path = None
    try:
        t0       = time.time()
        tts_path = text_to_speech_urdu(response["refined_answer"], verbose=verbose)
        tts_time = time.time() - t0
        print(f"\n🔊 Audio response saved: {tts_path}  ({tts_time:.1f}s)")
    except Exception as e:
        print(f"\n⚠️  TTS failed: {e}")
        tts_time = 0

    return {
        "input_mode"    : input_mode,
        "audio_path"    : audio_path,
        "transcribed_text": transcribed_text,
        "farmer_text"   : farmer_text,
        "enhanced_query": enhanced_query,
        "rag_results"   : rag_results,
        "good_results"  : good_results,
        "using_rag"     : using_rag,
        "raw_rag_answer": good_results[0]['answer'] if good_results else None,
        "final_answer"  : response["refined_answer"],
        "audio_response" : tts_path,
        "timing": {
            "stt"     : stt_time,
            "enhance" : enhance_time,
            "search"  : search_time,
            "generate": gen_time,
            "tts"     : tts_time,
            "total"   : total_time,
        },
    }

print("✅ run_full_pipeline() ready.")

✅ run_full_pipeline() ready.


## Step 10 — Test with Text Input
Test without needing an audio file.

In [52]:
# test_questions = [
#     "گندم میں زنگ کی بیماری کا کیا کرنا چاہیے؟",
#     "چاول میں کیڑے مار دوا کب دینی چاہیے؟",
#     "کپاس کو پانی کب دینا چاہیے؟",
#     "گندم میں یوریا کھاد کب ڈالنی چاہیے؟",
#     "مکئی کی فصل میں کھاد کی مقدار کیا ہے؟",
# ]

# # Change index to try a different question
# result = run_full_pipeline(text_input=test_questions[1], verbose=True)

In [53]:
# Example: voice input (audio file path)
audio_file = "../stt-finetune/training/asr_dataset/data/Recording.m4a"
result = run_full_pipeline(audio_path=audio_file, verbose=True)


############################################################
# AGRICULTURE ASSISTANT — FULL PIPELINE
############################################################
[Input mode: voice]
STT — SPEECH TO TEXT (Fine-tuned Whisper Urdu v2)
Audio: ../stt-finetune/training/asr_dataset/data/Recording.m4a

✅ Transcription: سر یہ کپاس پر میلی بگ کا اٹیک ہوا ہے بتا دی کیا کرنا ہے۔

[STT text]: سر یہ کپاس پر میلی بگ کا اٹیک ہوا ہے بتا دی کیا کرنا ہے۔

STEP 1 — LLM QUERY ENHANCEMENT (gemma2:9b)
  Original          : سر یہ کپاس پر میلی بگ کا اٹیک ہوا ہے بتا دی کیا کرنا ہے۔
  Detected language : Urdu
  Enhanced (English): Cotton plant is attacked by mealybug what to do
  Crop              : Cotton
  Topic             : Pest Management
  Stage             : Any
  Intent            : Recommendation
  Entity            : Mealybug
  Keywords          : cotton, mealybug, pest control, recommendation
  Translation conf  : 0.95
  Ambiguity notes   : None

RAG RESULTS (all matches + similarity scores)

  ✅ Res

---
## Quick Reference

| Function | What it does |
|---|---|
| `transcribe_audio(path)` | Audio → transcribed text (Whisper Urdu v2) |
| `enhance_farmer_query(text)` | LLM cleans + structures input before RAG |
| `rag_search(enhanced_query)` | ChromaDB vector search → top matches |
| `generate_farmer_response(...)` | LLM refines RAG answer → returns `{raw_rag_answer, refined_answer}` |
| `run_full_pipeline(audio_path=, text_input=)` | Full end-to-end |
| `interactive_mode()` | Loop for continuous Q&A |
| `run_evaluation()` | Benchmark + charts |

---

## VRAM on RTX 4060 Laptop (8GB)
| Component | VRAM |
|---|---|
| Gemma2:9b (Ollama) | ~5.5 GB |
| Fine-tuned Whisper | ~0.5 GB |
| SentenceTransformer | ~0.1 GB |
| **Total** | **~6.1 GB ✅** |